flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/ProcTimeBoundedProcessingOverProcessFunction.scala

```diff
@@ -0,0 +1,141 @@
+  * [[org.apache.flink.streaming.api.datastream.DataStream]]
+  *
+  * @param aggregates the list of all [[org.apache.flink.table.functions.AggregateFunction]]
+  *                   used for this aggregation
+  * @param aggFields  the position (in the input Row) of the input value for each aggregate
+  * @param forwardedFieldCount Is used to indicate fields in the current element to forward
+  * @param rowTypeInfo Is used to indicate the field schema
+  * @param time_boundary Is used to indicate the processing time boundaries
+  */
+class ProcTimeBoundedProcessingOverProcessFunction(
+    private val aggregates: Array[AggregateFunction[_]],
+    private val aggFields: Array[Int],
+    private val forwardedFieldCount: Int,
+    private val rowTypeInfo: RowTypeInfo,
+    private val time_boundary: Long)
+  extends ProcessFunction[Row, Row] {
+
+  Preconditions.checkNotNull(aggregates)
+  Preconditions.checkNotNull(aggFields)
+  Preconditions.checkArgument(aggregates.length == aggFields.length)
+
+  private var accumulators: Row = _
+  private var output: Row = _
+  private var windowBuffer: ListState[Tuple2[Long,Row]] = null
+  private var state: ValueState[Row] = _
+
+  
+  override def open(config: Configuration) {
+    output = new Row(forwardedFieldCount + aggregates.length)
+    
+    accumulators = new Row(aggregates.length)
+    var i = 0
+    while (i < aggregates.length) {
+        accumulators.setField(i, aggregates(i).createAccumulator())
+        i += 1
+      } 
+    
+    // We keep the elements received in a list state 
+    // together with the ingestion time in the operator
+    val bufferDescriptor: ListStateDescriptor[Tuple2[Long,Row]] = 
+    new ListStateDescriptor[Tuple2[Long,Row]]("windowBufferState", classOf[Tuple2[Long,Row]])
+    windowBuffer = getRuntimeContext.getListState(bufferDescriptor)
+
+    val stateDescriptor: ValueStateDescriptor[Row] =
+    new ValueStateDescriptor[Row]("overState", classOf[Row] , accumulators)      
+    state = getRuntimeContext.getState(stateDescriptor)
+  }
+
+  override def processElement(
+    input: Row,
+    ctx: ProcessFunction[Row, Row]#Context,
+    out: Collector[Row]): Unit = {
+
+    var current_time = System.currentTimeMillis()
+    //buffer the event incoming event
+    windowBuffer.add(new Tuple2(
+      current_time,
+      input))
+      
+    var i = 0
+
+    var accumulators = state.value()
+
+    //set the fields of the last event to carry on with the aggregates
+    i = 0
+    while (i < forwardedFieldCount) {
+      output.setField(i, input.getField(i))
+      i += 1
+    }
+
+     //update the elements to be removed and retract them from aggregators
+    var iter = windowBuffer.get.iterator()
+    var continue:Boolean = true
+    
+    while(continue && iter.hasNext())
+    {
+      var currentElement:Tuple2[Long,Row]= iter.next()  
+      if(currentElement._1<time_boundary){
+        iter.remove()
```

Comment:
Just chiming it because you tagged me earlier. 😃 

You correctly analysed that using a List state with `ProcessFunction` or a window would be equal since they both use the same underlying state underneath. `ProcessFunctin` allows a more elaborate approach, such as using `MapState`. Within that map state, would you still keep a whole window (per-key) in the list in the map state value? That could also be very expensive.

On a side note, you should be very aware that whatever you decide you probably have to stick with it for a while if you don't want to break savepoints that users created with one version and a given "state schema".